lib/iris/util.py

```diff
@@ -696,6 +696,27 @@ def _build_full_slice_given_keys(keys, ndim):
     return full_slice
 
 
+def _slice_data_with_keys(data, keys, n_parent_dims):
+    # Combine _build_full_slice_given_keys and column_slices_generator to index
+    # any array-like object with arbitrary keys using an orthogonal indexing
+    # interpretation.
+    # This means that both numpy and lazy data will index in the same way.
+    # Return a dimension map, and the indexed data.
+    #
+    # NOTE: by using slicing on only one index at a time, this also generally
+    # avoids copying the data, except when a key contains a list of indices.
+    full_slice = _build_full_slice_given_keys(keys, n_parent_dims)
+
+    dims_mapping, slices_iter = column_slices_generator(full_slice,
+                                                        n_parent_dims)
+    for this_slice in slices_iter:
+        data = data[this_slice]
+        if data.ndim > 0 and min(data.shape) < 1:
+            raise IndexError('Cannot index with zero length slice.')
```

Comment:
Existing code *allows* slicing a cube to 'nothing', like cube[2:2], but *disallows* this for a Coord (with the message here).  So..
```
>>> cube = iris.cube.Cube(np.arange(5))
>>> print cube[1:1]
unknown / (unknown)                 (-- : 0)
>>> cube.add_aux_coord(iris.coords.DimCoord(np.arange(5), long_name='x'), 0)
>>> print cube[1:1]
Traceback (most recent call last):
 . . .
  File "/opt/scitools/environments/experimental/2017_04_27/lib/python2.7/site-packages/iris/coords.py", line 492, in __getitem__
    raise IndexError('Cannot index with zero length '
IndexError: Cannot index with zero length slice.
>>> 
```

There are tests that check for this error, like [here](https://github.com/SciTools/iris/blob/3eb48980cb58a7784d9003d9fbe6f94003c1ce87/lib/iris/tests/test_cdm.py#L610)

So, I've retained it, but the restriction now applies to Cubes as well -- that didn't break any tests :crossed_fingers: 
This effectively disallows it for cubes as well,